### DISCUSSION TITLE
Loadable Node

### DISCUSSION THREAD

**Discussion by litewarp** [hayes/pothos#72] [Q&A] [answered]
I am having trouble with the [loadable node builder](https://giraphql.com/plugins/dataloader#using-with-the-relay-plugin). Specifically, I am confused about how to use classes to get isTypeOf checking using the Node Builder. If I try and pass the class to the builder (like with objects), I get the error that `Ref Class User { ... } has not been implemented`. If I try it with just the string 'User', the node gets created, but if I query by id, it returns the `Abstract type \"Node\" must resolve to an Object type at runtime for field \"Query.node\"` error.

Is there something obvious I am doing wrong? How should I implement this?

```JSX
export class User {
  id: string
  firstName: string
  lastName: string
  name: string
  email: string

  constructor(id: string, name: string, email: string, firstName: string, lastName: string) {
    this.id = id
    this.firstName = firstName
    this.lastName = lastName
    this.name = name
    this.email = email
  }
}

const UserNode = builder.loadableNode('User', {
  id: {
    resolve: (user) => user.id
  },
  isTypeOf: (obj: Record<string, unknown>) => obj instanceof User,
  fields: (t) => ({
    firstName: t.string({
      resolve: (src) => src.firstName
    }),
    lastName: t.string({
      resolve: (src) => src.lastName,
      nullable: true
    }),
    name: t.string({
      resolve: (src) => (src.lastName ? `${src.firstName} ${src.lastName}` : src.firstName)
    }),
    email: t.string({
      resolve: (src) => src.email
    }),
  }),
  load: async (ids: string[], context: ContextType): Promise<User[]> => {
    ...
  },
})

```

**Comment by hayes** [comment]
~This is likely a bug.  I'll try to fix it tonight~

**Reply by litewarp** [reply]
That would be amazing. Thanks.

**Comment by hayes** [comment]
I've tried to reproduce this, but it seems like everything is working as intended.  Can you add some console.logs into your isTypeOf handlers to see that it is being called, and that the check is returning true?

**Reply by litewarp** [reply]
The type is returning false for me. The object shape is correct but it's not a class. 

Maybe I am missing something, but besides the isTypeOf function, how does the node builder know that it is returning a "User" other than the string?

Do you have an implementation you could share? I didn't see a test in the plugin-relay package for the loadableNode.

**Reply by litewarp** [reply]
See screenshots below. The first is with the type defined as a class. I get the following typescript error and the "ref ... has not been implemented" error in the console. If I put in a string, I get the can't resolve Node type error in GraphiQL.

**Error**
<img width="688" alt="Screen Shot 2021-05-28 at 8 02 54 PM" src="https://user-images.githubusercontent.com/1900476/120056272-e1b8b680-bfef-11eb-8bff-797c0a0f0e87.png">

**With Class Passed In**
<img width="1510" alt="Screen Shot 2021-05-28 at 8 01 29 PM" src="https://user-images.githubusercontent.com/1900476/120056263-d5ccf480-bfef-11eb-83ab-4d26bd3bf734.png">

**With String Passed In**
<img width="1510" alt="Screen Shot 2021-05-28 at 8 00 34 PM" src="https://user-images.githubusercontent.com/1900476/120056256-c8b00580-bfef-11eb-8fbb-fdd9565b6093.png">

**Reply by hayes** [reply]
Interesting.  Sounds like an issue with the load function.  Can you confirm it is actually loading class instances?  Typescript let's you resolve objects of the right shape even if it's not actually an instance.

**Reply by hayes** [reply]
Using a class as the name is currently not supported, so the string version is correct

**Reply by litewarp** [reply]
Nevermind. I found the [test example](https://github.com/hayes/giraphql/blob/987b78e9195d1cbfc4c6f31e6b7f37cf8f869bb2/packages/plugin-dataloader/tests/example/schema/index.ts#L19) but it doesn't use classes. 

Is there no auto-magic resolution with classes when using the node interface? It's fine if I have to manually write a resolveType function, but it changes how I'll structure the underlying data. For example, I'd have to either pull in labels from the database or make sure each model from the database has a property unique to the model.

Sorry to bug you with so many questions. I appreciate your responsiveness.

**Reply by litewarp** [reply]
> Interesting.  Sounds like an issue with the load function.  Can you confirm it is actually loading class instances?  Typescript let's you resolve objects of the right shape even if it's not actually an instance.  

How would I figure out whether it's loading class instances?

**Reply by hayes** [reply]
You could save the return value as a variable and do an instance of check on one of the items in the returned list inside the load function.  Hard to tell without knowing more about what the load function does.

**Reply by litewarp** [reply]
Ahh I think I get it and why it's not working as I intended. 

My fetch function doesn't naturally return a class like Mongoose or TypeORM does. So I'd have to instantiate the class before returning it anyways.

Im not a big user of classes so in the past I'd just include a property on the returned fetch object that the isTypeOf resolver can check. 

Thanks again for your help.

**Reply by litewarp** [reply]
This was the answer. Posting it as a separate comment so I can mark it as resolved.

**Comment by litewarp** [comment, marked as answer]
The answer for anyone looking is that your load function has to return a class instance. So if your fetch call does not do it automatically, you need to return the class -- i.e. `return new User({ ...properties })`
